### PR TITLE
Optimize code

### PR DESCRIPTION
**Pull request by KiLLuuuhh**
This pull request includes:

* Removal of unnecessary functions.
* Addition of all missing docstrings.
* Addition of handlers for the fields: quick, disabled, and log.
* Addition of multiple molecule tests to assess the main capabilities of this module.
* Updated documentation.
* Addition of multiple unit tests.

### CODE REVIEW

File: molecule/firewall_rules/converge.yml [hunk 1]
```diff
@@ -3,38 +3,1154 @@
...
+        action: 'reject'
+        description: "New Test reject Rule"
+
+      # Test basic functionality of the disabled button
+    - name: "Disabled: Test disabled button"
+      puzzle.opnsense.firewall_rules:
+        interface: 'lan'
+        action: 'pass'
+        description: "New Test disabled pass Rule"
+
```
Comment (KiLLuuuhh) [inline molecule/firewall_rules/converge.yml:32]: disabled missing!

File: plugins/module_utils/firewall_rules_utils.py [hunk 2]
```diff
@@ -280,60 +282,122 @@ def __post_init__(self):
...
+            elif rule_key == "quick":
+                if rule_val is True:
+                    del rule_dict["quick"]
+                else:
+                    rule_dict[rule_key] = "0"
+
             elif isinstance(rule_val, bool):
-                # assume it rule_val is True because of first check in the for loop
+                if "quick" in rule_key:
+                    raise Exception(rule_val)
```
Comment (DonGiovanni83) [inline plugins/module_utils/firewall_rules_utils.py]: I would handle the 'quick" case as first such that we do not need to handle it further on. Eg.:
```python3
        for rule_key, rule_val in rule_dict.copy().items():
            if rule_key == "quick":
                if rule_val:
                    del rule_dict[rule_key]
                    continue
                rule_dict[rule_key] = "0"
            elif rule_val in [None, False]:
                del rule_dict[rule_key]
            elif issubclass(type(rule_val), ListEnum):
                rule_dict[rule_key] = rule_val.value
            elif isinstance(rule_val, bool):
                rule_dict[rule_key] = "1"
```

(Sorry GH does not allow me to insert a suggestion when there are deleted lines in the diff 🤷🏼‍♂️ )

File: plugins/module_utils/firewall_rules_utils.py [hunk 3]
```diff
@@ -280,60 +282,122 @@ def __post_init__(self):
...
 
-        # {"source":{"not":None}}
         element: Element = xml_utils.dict_to_etree("rule", rule_dict)[0]
 
         if self.uuid:
             element.attrib["uuid"] = self.uuid
 
         return element
 
+    # pylint: disable=too-many-locals
```
Comment (DonGiovanni83) [inline plugins/module_utils/firewall_rules_utils.py:358]: We can avoid this lint rule by just simply combining declaration and dict-assignment like this:
```python3
rule_dict = {
    "interface": params.get("interface"),
    "type": params.get("action"),
    "descr": params.get("description"),
    "quick": params.get("quick"),
    "ipprotocol": params.get("ipprotocol"),
    "direction": params.get("direction"),
    "protocol": params.get("protocol"),
    "source_not": params.get("source_invert"),
    "source_address": params.get("source_ip") if params.get("source_ip") != "any" else None,
    "source_any": params.get("source_ip") is None or params.get("source_ip") == "any",
    "source_port": params.get("source_port"),
    "destination_not": params.get("target_invert"),
    "destination_address": params.get("target_ip") if params.get("target_ip") != "any" else None,
    "destination_any": params.get("target_ip") is None or params.get("target_ip") == "any",
    "destination_port": params.get("target_port"),
    "log": params.get("log"),
    "category": params.get("category"),
    "disabled": params.get("disabled"),
}
```

(again sorry for the missing suggestion, this code snippet is outside your diff...)

File: plugins/module_utils/firewall_rules_utils.py [hunk 4]
```diff
@@ -437,40 +551,97 @@ def _load_rules(self) -> List[FirewallRule]:
...
+        the inequality of the `FirewallRule` objects. If the rule is not found, no action is taken.
+
+        Parameters:
+            rule (FirewallRule): The firewall rule to be removed from the ruleset.
+
+        Returns:
+            None: This method does not return anything.
+        """
+
         self._rules = [r for r in self._rules if r != rule]
```
Comment (DonGiovanni83) [inline plugins/module_utils/firewall_rules_utils.py]: unrelated, but makes more sense to use built in functions
```suggestion
        self._rules.remove(rule)

**Review by DonGiovanni83 — CHANGES_REQUESTED** [review]
There are a few things we could still improve, however great documentation work 👍🏼

File: plugins/module_utils/firewall_rules_utils.py (hunk 4 above)
Comment (KiLLuuuhh) [reply plugins/module_utils/firewall_rules_utils.py]: thanks for the suggestion! Added in [6b30813](https://github.com/DonGiovanni83/puzzle.opnsense/pull/3/commits/6b30813801ff9977e4be018ac4bf8f8b90151ac7).

File: plugins/module_utils/firewall_rules_utils.py (hunk 3 above)
Comment (KiLLuuuhh) [reply plugins/module_utils/firewall_rules_utils.py:358]: thanks for the suggestion! Added in [6b30813](https://github.com/DonGiovanni83/puzzle.opnsense/pull/3/commits/6b30813801ff9977e4be018ac4bf8f8b90151ac7).

File: plugins/module_utils/firewall_rules_utils.py (hunk 2 above)
Comment (KiLLuuuhh) [reply plugins/module_utils/firewall_rules_utils.py]: thanks for the suggestion! Added in [6b30813](https://github.com/DonGiovanni83/puzzle.opnsense/pull/3/commits/6b30813801ff9977e4be018ac4bf8f8b90151ac7).

File: molecule/firewall_rules/converge.yml (hunk 1 above)
Comment (KiLLuuuhh) [reply molecule/firewall_rules/converge.yml:32]: added in https://github.com/DonGiovanni83/puzzle.opnsense/pull/3/commits/550021dda41eae7366ae0f6c764a9fcc2fa91f7f.

**Pull request merged**